File: plugins/cwallmonitor/server/py/src/cwm_mcp/registry/crypto.py
"""AES for pending config blobs. See ../compat/SECURITY.md.

Legacy wire = AES-256-CTR (16-byte IV, no auth tag). New wire (firmware
>= PENDING_GCM_MIN_FW) = AES-256-GCM: 12-byte nonce, payload = ct||tag,
AAD = ASCII decimal of pending.version. See ../compat/vectors/aes_gcm.json.
"""

from __future__ import annotations

import os

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

PENDING_NONCE_LEN = 16  # AES block size (CTR IV)
PENDING_GCM_NONCE_LEN = 12  # AES-GCM nonce
# Firmware at or above this version carries the GCM decrypt path. The broker
# only emits "enc":"gcm" when the live X-Cwm-Fw-Version says so.
PENDING_GCM_MIN_FW = "0.9.0"


def encrypt_pending(key: bytes, plaintext: bytes) -> tuple[bytes, bytes]:
    if len(key) != 32:
        raise ValueError(f"registry/crypto: key must be 32 bytes, got {len(key)}")
    nonce = os.urandom(PENDING_NONCE_LEN)
    cipher = Cipher(algorithms.AES(key), modes.CTR(nonce))
    enc = cipher.encryptor()
    ciphertext = enc.update(plaintext) + enc.finalize()
    return nonce, ciphertext


def decrypt_pending(key: bytes, nonce: bytes, ciphertext: bytes) -> bytes:
    if len(key) != 32:
        raise ValueError(f"registry/crypto: key must be 32 bytes, got {len(key)}")
    if len(nonce) != PENDING_NONCE_LEN:
        raise ValueError(f"registry/crypto: nonce must be {PENDING_NONCE_LEN} bytes, got {len(nonce)}")
    if not ciphertext:
        raise ValueError("registry/crypto: empty ciphertext")
    cipher = Cipher(algorithms.AES(key), modes.CTR(nonce))
    dec = cipher.decryptor()
    return dec.update(ciphertext) + dec.finalize()


def encrypt_pending_gcm(
    key: bytes, plaintext: bytes, version: int, nonce: bytes | None = None
) -> tuple[bytes, bytes]:
    """AES-256-GCM encrypt. Returns (nonce, ct||tag).

    AAD = ASCII decimal of `version` (e.g. 7 -> b"7"), binding the blob to
    pending.version. AESGCM.encrypt natively appends the 16-byte tag. nonce
    is for tests only; production passes None for a fresh 12-byte nonce.
    """
    if len(key) != 32:
        raise ValueError(f"registry/crypto: key must be 32 bytes, got {len(key)}")
    if nonce is None:
        nonce = os.urandom(PENDING_GCM_NONCE_LEN)
    elif len(nonce) != PENDING_GCM_NONCE_LEN:
        raise ValueError(
            f"registry/crypto: gcm nonce must be {PENDING_GCM_NONCE_LEN} bytes, got {len(nonce)}"
        )
    aad = str(version).encode("ascii")
    ct_tag = AESGCM(key).encrypt(nonce, plaintext, aad)
    return nonce, ct_tag


def decrypt_pending_gcm(
    key: bytes, nonce: bytes, ciphertext: bytes, version: int
) -> bytes:
    """AES-256-GCM decrypt of ct||tag. Raises on auth failure / bad lengths.

    Rejects non-32-byte keys and non-12-byte nonces (blocks the CTR-IV
    downgrade-strip described in aes_gcm.json's negative vectors)."""
    if len(key) != 32:
        raise ValueError(f"registry/crypto: key must be 32 bytes, got {len(key)}")
    if len(nonce) != PENDING_GCM_NONCE_LEN:
        raise ValueError(
            f"registry/crypto: gcm nonce must be {PENDING_GCM_NONCE_LEN} bytes, got {len(nonce)}"
        )
    if not ciphertext:
        raise ValueError("registry/crypto: empty ciphertext")
    aad = str(version).encode("ascii")
    return AESGCM(key).decrypt(nonce, ciphertext, aad)


def gcm_fw_gate_open(fw_version: str) -> bool:
    """True when the live X-Cwm-Fw-Version warrants GCM emission, i.e. its
    numeric MAJOR.MINOR.PATCH prefix is >= PENDING_GCM_MIN_FW.

    The -suffix is ignored on purpose: a dev build "0.9.0-dev.<ts>" carries
    the same decrypt code as the "0.9.0" release, so it must also get GCM.
    Garbage / empty / absent versions (pack returns None) fall back to CTR —
    we never gate on registry state, only on what the device reports."""
    from ..ota import pack_semver  # lazy: avoid import cycle (ota → registry)

    packed = pack_semver(fw_version)
    if packed is None:
        return False
    floor = pack_semver(PENDING_GCM_MIN_FW)
    assert floor is not None  # constant is well-formed
    return packed >= floor

File: plugins/cwallmonitor/server/js/src/registry/crypto.js
// Pending-blob encryption. Two wire schemes:
//   - legacy AES-256-CTR, 16-byte IV (no "enc" field) — pre-0.9.0 firmware.
//   - AES-256-GCM, 12-byte nonce, 16-byte tag appended to ciphertext, AAD =
//     ASCII decimal of pending.version ("enc":"gcm") — firmware >= 0.9.0.
// Mirrors registry/crypto.go. See ../compat/SECURITY.md and
// ../compat/vectors/aes_gcm.json (the byte-exact GCM contract).

import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";
import { packSemver } from "../ota.js";

export const PENDING_NONCE_LEN = 16;
export const PENDING_GCM_NONCE_LEN = 12;
export const PENDING_GCM_TAG_LEN = 16;

// Firmware at or above this version understands the GCM wire format. The
// broker only emits "enc":"gcm" when the live X-Cwm-Fw-Version request header
// is >= this; older firmware keeps getting the legacy CTR blob. Must equal
// compat/vectors/aes_gcm.json min_fw_version (asserted in the tests).
export const PENDING_GCM_MIN_FW = "0.9.0";

export function encryptPending(key, plaintext) {
  if (key.length !== 32) throw new Error(`registry/crypto: key must be 32 bytes, got ${key.length}`);
  const nonce = randomBytes(PENDING_NONCE_LEN);
  const cipher = createCipheriv("aes-256-ctr", key, nonce);
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  return { nonce, ciphertext: ct };
}

export function decryptPending(key, nonce, ciphertext) {
  if (key.length !== 32) throw new Error(`registry/crypto: key must be 32 bytes, got ${key.length}`);
  if (nonce.length !== PENDING_NONCE_LEN) throw new Error(`registry/crypto: nonce must be ${PENDING_NONCE_LEN} bytes, got ${nonce.length}`);
  if (!ciphertext || ciphertext.length === 0) throw new Error("registry/crypto: empty ciphertext");
  const dec = createDecipheriv("aes-256-ctr", key, nonce);
  return Buffer.concat([dec.update(ciphertext), dec.final()]);
}

// encryptPendingGCM encrypts the pending payload with AES-256-GCM. The 12-byte
// nonce is generated fresh unless `fixedNonce` is supplied (tests inject a
// pinned nonce to reproduce vector ciphertext byte-for-byte). The returned
// ciphertext is ct||tag — the 16-byte GCM tag appended, matching the native
// PSA / Go crypto/cipher GCM / Python AESGCM byte order. AAD binds the blob to
// pending.version: the ASCII decimal of `version` (e.g. 7 => the byte "7"),
// NOT the raw 4-byte integer.
export function encryptPendingGCM(key, plaintext, version, fixedNonce) {
  if (key.length !== 32) throw new Error(`registry/crypto: key must be 32 bytes, got ${key.length}`);
  let nonce;
  if (fixedNonce !== undefined) {
    if (fixedNonce.length !== PENDING_GCM_NONCE_LEN) {
      throw new Error(`registry/crypto: gcm nonce must be ${PENDING_GCM_NONCE_LEN} bytes, got ${fixedNonce.length}`);
    }
    nonce = fixedNonce;
  } else {
    nonce = randomBytes(PENDING_GCM_NONCE_LEN);
  }
  const cipher = createCipheriv("aes-256-gcm", key, nonce, { authTagLength: PENDING_GCM_TAG_LEN });
  const aad = Buffer.from(String(version >>> 0), "ascii");
  cipher.setAAD(aad);
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  return { nonce, ciphertext: Buffer.concat([ct, tag]) };
}

// decryptPendingGCM is the inverse of encryptPendingGCM: `ciphertext` is the
// ct||tag concatenation as it travels on the wire. Throws on a bad tag, wrong
// AAD (version mismatch), wrong key length, or a non-12-byte nonce — never
// returns plaintext on authentication failure.
export function decryptPendingGCM(key, nonce, ciphertext, version) {
  if (key.length !== 32) throw new Error(`registry/crypto: key must be 32 bytes, got ${key.length}`);
  if (nonce.length !== PENDING_GCM_NONCE_LEN) throw new Error(`registry/crypto: gcm nonce must be ${PENDING_GCM_NONCE_LEN} bytes, got ${nonce.length}`);
  if (!ciphertext || ciphertext.length < PENDING_GCM_TAG_LEN) throw new Error("registry/crypto: ciphertext shorter than gcm tag");
  const ct = ciphertext.subarray(0, ciphertext.length - PENDING_GCM_TAG_LEN);
  const tag = ciphertext.subarray(ciphertext.length - PENDING_GCM_TAG_LEN);
  const dec = createDecipheriv("aes-256-gcm", key, nonce, { authTagLength: PENDING_GCM_TAG_LEN });
  const aad = Buffer.from(String(version >>> 0), "ascii");
  dec.setAAD(aad);
  dec.setAuthTag(tag);
  return Buffer.concat([dec.update(ct), dec.final()]);
}

// gcmFwGate reports whether a live X-Cwm-Fw-Version header value is at or above
// PENDING_GCM_MIN_FW, i.e. the device understands the GCM wire format. Uses the
// SAME strict packSemver() the Go (ota.PackSemver / fwSupportsGCM) and Python
// (ota.pack_semver / gcm_fw_gate_open) brokers use, so identical headers gate
// identically across all three impls: the version must be a strict numeric
// MAJOR.MINOR.PATCH (exactly three components, ASCII digits, no leading zeros,
// in 8.8.16 range). An optional "-dev.<ts>" prerelease suffix is IGNORED (a
// 0.9.0-dev build carries the GCM decryptor). ANYTHING else — "0.9", "v0.9.0",
// "0.9.0+build", "00.9.0", out-of-range, empty/absent — fails packSemver and
// falls back to legacy CTR.
export function gcmFwGate(fwHeader) {
  const got = packSemver(fwHeader);
  if (got === null) return false;
  const floor = packSemver(PENDING_GCM_MIN_FW);
  return got >= floor;
}
